lib/psych/visitors/yaml_tree.rb

```diff
@@ -310,19 +310,20 @@ def visit_String o
           style = Nodes::Scalar::LITERAL
           plain = false
           quote = false
-        elsif o =~ /\n/
-          style = Nodes::Scalar::LITERAL
         elsif o == '<<'
           style = Nodes::Scalar::SINGLE_QUOTED
           tag   = 'tag:yaml.org,2002:str'
           plain = false
           quote = false
+        elsif o =~ /\n[^\Z]/  # match \n except blank line at the end of string
+          style = Nodes::Scalar::LITERAL
+        elsif o.length > line_width
```

Comment:
Can you do the same benchmark with [benchmark-ips](https://github.com/evanphx/benchmark-ips)?  Also note that we could refactor to a simple method call rather than a lambda:

```diff
diff --git a/lib/psych/visitors/yaml_tree.rb b/lib/psych/visitors/yaml_tree.rb
index aa7f10e..98b708d 100644
--- a/lib/psych/visitors/yaml_tree.rb
+++ b/lib/psych/visitors/yaml_tree.rb
@@ -60,6 +60,18 @@ module Psych
         create emitter, ss
       end
 
+      module Empty
+        def self.call o
+          false
+        end
+      end
+
+      class NonEmpty < Struct.new(:len)
+        def call o
+          o.length > len
+        end
+      end
+
       def initialize emitter, ss, options
         super()
         @started  = false
@@ -79,6 +91,10 @@ module Psych
 
           h[klass] = method
         end
+
+        @len_compare = options[:line_width] ?
+                        NonEmpty.new(options[:line_width]) :
+                        Empty
       end
 
       def start encoding = Nodes::Stream::UTF8
@@ -317,7 +333,7 @@ module Psych
           quote = false
         elsif o =~ /\n[^\Z]/  # match \n except blank line at the end of string
           style = Nodes::Scalar::LITERAL
-        elsif o.length > line_width
+        elsif @len_compare.call o
           style = Nodes::Scalar::FOLDED
           o << "\n" unless o =~ /\n\Z/  # to avoid non-default chomping indicator
         elsif o =~ /^[^[:word:]][^"]*$/
@@ -590,10 +606,6 @@ module Psych
           accept target.instance_variable_get(iv)
         end
       end
-
-      def line_width
-        @options[:line_width] || Float::INFINITY
-      end
     end
   end
 end
```

So you may want to add a test that calls a method that just returns `false`.